rbac/common/manager/rbac_manager.py

```diff
@@ -13,11 +13,90 @@
 # limitations under the License.
 # -----------------------------------------------------------------------------
 
+import logging
+import inspect
+
+from sawtooth_sdk.processor.exceptions import InvalidTransaction
+from rbac.common.protobuf.rbac_payload_pb2 import RBACPayload
 from rbac.common.user.user_manager import UserManager
 from rbac.common.role.role_manager import RoleManager
+from rbac.common.task.task_manager import TaskManager
+
+LOGGER = logging.getLogger(__name__)
+
+
+def walk_filter(value):
+    return str(type(value)).startswith("<class 'rbac.common.")
+
+
+def hasgetattr(value, attrib):
+    if hasattr(value, attrib):
+        return getattr(value, attrib)
+    return None
+
+
+def message_type_name(message_type):
+    items = RBACPayload.MessageType.DESCRIPTOR.values_by_name.items()
+    # return items[message_type][0] # breaks on OSX and/or Python 3.7
+    for (key, value) in items:
+        if value == message_type:
+            return key
+    return None
 
 
 class RBACManager:
     def __init__(self):
         self.user = UserManager()
         self.role = RoleManager()
+        self.task = TaskManager()
+        self.handlers = {}
+        self.walk_handlers(self, "rbac")
+
+    def walk_handlers(self, node, path):
+        props = [
+            name
+            for (name, value) in inspect.getmembers(node, walk_filter)
+            if not name.startswith("__")
+        ]
+        for prop in props:
+            new_path = path + "." + prop
+            value = getattr(node, prop)
+            message_type = hasgetattr(value, "message_type")
+            apply = hasgetattr(value, "apply")
+            if message_type is not None:
+                self.register_handler(new_path, message_type, apply)
+            self.walk_handlers(value, new_path)
+
+    def register_handler(self, path, message_type, apply):
+        if message_type in self.handlers:
+            raise KeyError(
+                "{} tried to register message {} that was already registered".format(
+                    path, message_type_name(message_type)
+                )
+            )
+
+        if apply is None:
+            return  # LOGGER.warning('%s message has no apply handler at %s', message_type_name(message_type), path)
```

Comment:
Please remove or uncomment the log statement.

Reply:
resolved